### PR TITLE
PES-2243: example tests

### PR DESCRIPTION
**Pull request by jan-potuznik-z**

### CODE REVIEW

File: src/Packetery/Module/Framework/WcCartTrait.php [hunk 1]
```diff
@@ -27,10 +27,11 @@ public function cartGetCartContents(): array {
...
 	/**
 	 * Gets cart total. This is the total of items in the cart, but after discounts. Subtotal is before discounts.
+	 * Really returns string, at least sometimes.
 	 *
-	 * @return float
+	 * @return float|string
 	 */
-	public function cartGetCartContentsTotal(): float {
+	public function cartGetCartContentsTotal() {
 		return WC()->cart->get_cart_contents_total();
```
Comment (mzk) [inline src/Packetery/Module/Framework/WcCartTrait.php:35]: what's about `return (float) WC()->cart->get_cart_contents_total();` ?

File: src/Packetery/Core/Api/Soap/Client.php [hunk 2]
```diff
@@ -9,8 +9,6 @@
 
 namespace Packetery\Core\Api\Soap;
 
-use Packetery\Core\Api\Soap\Request;
-use Packetery\Core\Api\Soap\Response;
```
Comment (mzk) [inline src/Packetery/Core/Api/Soap/Client.php:13]: why this wasn't covered by phcps?

File: tests/Module/CheckoutTest.php [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+	): void {
+		$wcProduct = $this->createMock( \WC_Product::class );
+		$wcProduct
+			->method( 'get_price' )
+			->willReturn( '100.00' );
+		$wcProduct
+			->method( 'get_category_ids' )
+			->willReturn( [ 1 ] );
+
+		$cartItem = [
```
Comment (mzk) [inline tests/Module/CheckoutTest.php:426]: Please keep "Create Variable" on line 426 and "Use the variable" on line 476 as soon as possible.
Screenshot 2024-08-21 at 11 11 12

File: src/Packetery/Module/Framework/WcCartTrait.php (hunk 1 above)
Comment (jan-potuznik-z) [reply src/Packetery/Module/Framework/WcCartTrait.php:35]: In my opinion adapter should not perform any logic and plugin should get raw values from the framework.

File: tests/Module/CheckoutTest.php [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+			->method( 'getDisallowedShippingRateIds' )
+			->willReturn( array_merge( [ Carrier\OptionPrefixer::getOptionId( 'anyDisallowedOnProduct' ) ], $productDisallowedRateIds ) );
+
+		$cart = $this->createMock( \WC_Cart::class );
+		$cart
+			->method( 'get_coupons' )
+			->willReturn( [] );
+
+		$wpAdapter                    = $this->getPacketeryMockFactory()->createWpAdapter();
+		$wcAdapter                    = $this->createMock( WcAdapter::class );
```
Comment (mzk) [inline tests/Module/CheckoutTest.php:449]: `$wcAdapter` - a) create variable on line 449, c) usage on line 459 and b) definition of lines 468-513?

I like to a) create variable, b) definition of mock and c) use the variable.

File: src/Packetery/Core/Api/Soap/Client.php (hunk 2 above)
Comment (jan-potuznik-z) [reply src/Packetery/Core/Api/Soap/Client.php:13]: Very good idea, is it possible to include in this PR, or should we add it together with other missing useful rules?

File: tests/Module/CheckoutTest.php [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+			->expects( self::atLeast( $expectedRateCount ) )
+			->method( 'applyFilters' );
+
+		$wpAdapter
+			->method( 'didAction' )
+			->willReturn( 1 );
+
+		$rates = $checkout->getShippingRates( $allowedCarrierNames );
+
+		self::assertCount( $expectedRateCount, $rates );
```
Comment (mzk) [inline tests/Module/CheckoutTest.php:538]: assert count & values pls.

File: tests/Module/WithMockFactory.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare( strict_types=1 );
+
+namespace Tests\Module;
+
+trait WithMockFactory {
+
+	public function getPacketeryMockFactory(): MockFactory {
```
Comment (mzk) [inline tests/Module/WithMockFactory.php:9]: what is the purpose of this trait?

File: tests/Module/RateCalculatorTest.php [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+			[ null, $defaultCarrierOptions, 100, 31, false ],
+			[ 111, $carrierOptionsProductValuePricing, 100, 1, false ],
+			[ 222, $carrierOptionsProductValuePricing, 150, 15, false ],
+			[ 222, $carrierOptionsProductValuePricing, 200, 20, false ],
+			[ 333, $carrierOptionsProductValuePricing, 300, 25, false ],
+			[ null, $carrierOptionsProductValuePricing, 400, 31, false ],
+			'free shipping threshold must be reached'     => [
+				0,
+				$defaultCarrierOptions,
+				15000,
```
Comment (mzk) [inline tests/Module/RateCalculatorTest.php:101]: Consider to use array keys to improving readability - I don't know what is 15000 value, 25 and false.
`'totalProductValue' => 15000`

File: tests/Module/RateCalculatorTest.php [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+			[ 22, $defaultCarrierOptions, 100, 20, false ],
+			[ 33, $defaultCarrierOptions, 100, 25, false ],
+			[ null, $defaultCarrierOptions, 100, 31, false ],
+			[ 111, $carrierOptionsProductValuePricing, 100, 1, false ],
+			[ 222, $carrierOptionsProductValuePricing, 150, 15, false ],
+			[ 222, $carrierOptionsProductValuePricing, 200, 20, false ],
+			[ 333, $carrierOptionsProductValuePricing, 300, 25, false ],
+			[ null, $carrierOptionsProductValuePricing, 400, 31, false ],
+			'free shipping threshold must be reached'     => [
+				0,
```
Comment (mzk) [inline tests/Module/RateCalculatorTest.php:99]: expectedCost => 0

File: tests/Module/MockFactory.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+use Packetery\Module\CurrencySwitcherFacade;
+use Packetery\Module\Framework\WpAdapter;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+class MockFactory {
+	private TestCase $testCase;
+
+	public function __construct( TestCase $testCase ) {
+		$this->testCase = $testCase;
```
Comment (mzk) [inline tests/Module/MockFactory.php:16]: I don't like it.

File: tests/Module/MockFactory.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	}
+
+	public function createWpAdapter(): WpAdapter|MockObject {
```
Comment (mzk) [inline tests/Module/MockFactory.php:19]: `public function createWpAdapter( TestCase $testCase): WpAdapter|MockObject {`
or use BaseTestCase

**Review by mzk — CHANGES_REQUESTED** [review]

File: tests/Module/WithMockFactory.php (hunk 6 above)
Comment (jan-potuznik-z) [reply tests/Module/WithMockFactory.php:9]: That is question for Filip.

File: tests/Module/MockFactory.php (hunk 9 above)
Comment (mzk) [reply tests/Module/MockFactory.php]: The reason is in my next review comment few lines under https://github.com/Zasilkovna/WooCommerce/pull/589#discussion_r1724743523

File: src/Packetery/Core/Api/Soap/Client.php (hunk 2 above)
Comment (mzk) [reply src/Packetery/Core/Api/Soap/Client.php:13]: please in the new MR.

File: src/Packetery/Module/Framework/WcCartTrait.php (hunk 1 above)
Comment (mzk) [reply src/Packetery/Module/Framework/WcCartTrait.php:35]: ok, up to you.

Comment (mzk) [reply src/Packetery/Module/Framework/WcCartTrait.php:35]: I'm voting for adding `(float)` => adapter can validate data, improve readability, phpdoc comments, native types etc.

File: tests/Module/RateCalculatorTest.php [hunk 11]
```diff
@@ -130,18 +120,14 @@ public static function calculationDataProvider(): array {
 	 * @dataProvider calculationDataProvider
 	 */
 	public function testGetShippingRateCost( ?float $expectedCost, Options $carrierOptions, float $totalProductValue, float $cartWeight, bool $isCouponApplied ): void {
-		$wpAdapter = $this->getPacketeryMockFactory()->createWpAdapter();
+		$mockFactory = new MockFactory();
+		$wpAdapter   = $mockFactory->createWpAdapter( $this );
```
Comment (mzk) [inline tests/Module/RateCalculatorTest.php:124]: mockFactory::createWpAdapter can be static

File: tests/Module/RateCalculatorTest.php [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => true,
+			],
+		];
+	}
+
+	/**
+	 * @dataProvider calculationDataProvider
+	 */
+	public function testGetShippingRateCost( ?float $expectedCost, Options $carrierOptions, float $totalProductValue, float $cartWeight, bool $isCouponApplied ): void {
```
Comment (mzk) [inline tests/Module/RateCalculatorTest.php:122]: the parameter names of this method should match the dataProvider keys.
See `$cartWeight` vs `cartWeightKg`

File: src/Packetery/Core/Api/Soap/Client.php (hunk 2 above)
Comment (jan-potuznik-z) [reply src/Packetery/Core/Api/Soap/Client.php:13]: Ticket is on the way.

File: tests/Module/CheckoutTest.php (hunk 5 above)
Comment (jan-potuznik-z) [reply tests/Module/CheckoutTest.php:534]: Tested method should be refactored soon. After that, this change would not be meaningful.

**Review by mzk — APPROVED** [review]

**Pull request merged**